pandas/core/groupby/groupby.py

```diff
@@ -689,7 +689,22 @@ def _set_result_index_ordered(self, result):
             result.set_axis(index, axis=self.axis, inplace=True)
             result = result.sort_index(axis=self.axis)
 
-        result.set_axis(self.obj._get_axis(self.axis), axis=self.axis, inplace=True)
+        return self._restore_grouper_index(result)
+
+    def _restore_grouper_index(self, result: FrameOrSeries) -> FrameOrSeries:
+        # GH 35612
+        # result.index is a standard index => may need to restore original index
+
+        rows_dropped: bool = len(result) < len(self._selected_obj)
+        # TODO: address case of axis==1, dropped columns
+
+        if self.dropna and not self.axis and rows_dropped:
```

Comment:
let's be clear: `self.axis == 0`

Reply:
why does rows dropped actually matter here?

Reply:
The rows_dropped check allows us to avoid an indexing operation so I am okay with it; but also confused why tests would fail without it. What tests fail?

What is currently called "original_index" is not the original index, namely it is missing entries. Perhaps "labels" would be better?

Recommend using the if/else blocks to set labels and then make a single call to self.set_index rather than setting the index from two different places.

Reply:
I was looking at some old commits; ignore the above.